src/lib/modules/scaffolding/index.ts

```diff
@@ -0,0 +1,60 @@
+import { Contract } from "../../../typings/contract";
+import { Embark } from "../../../typings/embark";
+import { Solidity } from "./contractLanguage/solidity";
+import { Description } from "./description";
+import { React } from "./framework/react";
+import { ContractLanguage, Framework, Options } from "./options";
+
+export default class Scaffolding {
+
+  constructor(private embark: Embark, private options: any) {
+    this.embark.events.setCommandHandler("scaffolding:generate:contract", this.generateContract.bind(this));
+    this.embark.events.setCommandHandler("scaffolding:generate:ui", this.generateUi.bind(this));
+  }
+
+  private readonly contractLanguageStrategy = (description: Description, options: Options) => {
+    switch (options.contractLanguage) {
+      case ContractLanguage.Solidity: {
+        return new Solidity(this.embark, description, options);
+      }
+    }
+  }
+
+  private readonly frameworkStrategy = (description: Description, contracts: Contract[], options: Options) => {
+    switch (options.framework) {
+      case Framework.React: {
+        return new React(this.embark, description, contracts, options);
+      }
+    }
+  }
+
+  private readonly parseAndValidate = (cmdLineOptions: any) => {
+    const options = new Options(this.embark.logger, cmdLineOptions.framework, cmdLineOptions.contractLanguage, cmdLineOptions.overwrite);
+    const description = new Description(this.embark.logger, cmdLineOptions.contractOrFile, cmdLineOptions.fields);
+    options.validate();
+    description.validate();
+
+    return {description, options};
+  }
+
+  private readonly generateContract = async (cmdLineOptions: any, cb: (files: string[]) => void) => {
+    const {description, options} = this.parseAndValidate(cmdLineOptions);
+    const files = await this.contractLanguageStrategy(description, options).build();
+    cb(files);
```

Comment:
Since this is brand new code, I'd say we take the chance and don't introduce additional APIs that receive callbacks and instead *at least* return a promise, so it's easier to compose these methods.

Reply:
In order to communicate with the rest of embark, you need Callback.
There is 2 function that communicates with embark:
generateContract and generateUI, they are the only 2 using callbacks